text/0000-serialization-format.md

```diff
@@ -0,0 +1,116 @@
+# Serialization Format
+
+## Summary
+
+[summary]: #summary
+
+Definition of a versioned, backwards-compatible and extensible serialization
+format for the state of Cartographer.
+
+This RFC is a small deviation of the previously proposed header format from
+[RFC-0015](https://github.com/googleCartographer/rfc/text/0015-serialization-header.md).
+
+## Motivation
+
+[motivation]: #motivation
+
+At the moment, the file format for serializing the state of Cartographer is
+defined purely within the serialization & deserialization routines of
+Cartographer. Since we serialize all information into a binary stream of
+`protobuf` messages (*pbstream*) without a format-version information, extending
+or changing the format inevitably leads to a breaking change for Cartographer
+users, as there is no way of falling back to load previously serialized states.
+
+This RFC is going to address this issue by introducing a `SerializationHeader`
+message and an interface for parsing serialized states, independent of the
+file-format.
+
+## Approach
+
+[approach]: #approach
+
+For being able to identify the underlying file-format, we introduce a
+`SerializationHeader` message.
+
+```
+message SerializationHeader {
+  string format_version = 1;
+}
+```
+
+We further change the current `SerializedData` protobuf message, which is meant
+to be able to contain `oneof` any of the message types required for serializing
+the mapping state:
+
+```
+message SerializedData {
+  oneof data {
+    PoseGraph pose_graph = 1;
+    AllTrajectoryBuilderOptions all_trajectory_builder_options = 2;
+    Submap submap = 3;
+    Node node = 4;
+    ImuData imu_data = 5;
+    OdometryData odometry_data = 6;
+    FixedFramePoseData fixed_frame_pose_data = 7;
+    TrajectoryData trajectory_data = 8;
+    LandmarkData landmark_data = 9;
+  }
+}
+```
+
+For backwards compatibility, we will rename the original `SerializedData`
+definition to `LegacySerializedData`.
+
+By keeping all the payload related data wrapped in a single message type, we can
+also easily switch to a different underlying storage format in the future (e.g.
+[Riegeli](https://github.com/google/riegeli/)).
+
+### Generic File Format
+
+Using the above messages, the generic file format is defined as:
+
+*Generic File Format* |
```

Comment:
Please mention that we put magic bytes "0x7b1d1f7b5bf501db" in front and that all messages are gzipped (each message separately), as in `ProtoStreamWriter`.